torch/distributed/fsdp/_shard_utils.py

```diff
@@ -73,6 +75,21 @@ def _gather_state_dict(
     return new_state_dict
 
 
+def _gather_dtensor_state_dict(
```

Comment:
Should we combine `_gather_state_dict` and `_gather_dtensor_state_dict`? I don't see any issue of merge `_gather_dtensor_state_dict` into `_gather_state_dict`. It will also support the case where a state_dict have both ShardedTensor and DTensor.

Reply:
I am separating this since the goal is to deprecate sharded_tensors eventually. So once we turn dtensor to default, we can simply just delete the _gather_state_dict and keep _gather_dtensor_state_dict. What do you think?